xcm/src/v1/multilocation.rs

```diff
@@ -36,83 +39,385 @@ use parity_scale_codec::{self, Decode, Encode};
 ///
-/// This specific `MultiLocation` implementation uses a Rust `enum` in order to make pattern matching easier.
-///
-/// The `MultiLocation` value of `Here` simply refers to the interpreting consensus system.
-#[derive(Clone, Eq, PartialEq, Ord, PartialOrd, Encode, Decode, Debug)]
-pub enum MultiLocation {
-	/// The interpreting consensus system.
-	Here,
-	/// A relative path comprising 1 junction.
-	X1(Junction),
-	/// A relative path comprising 2 junctions.
-	X2(Junction, Junction),
-	/// A relative path comprising 3 junctions.
-	X3(Junction, Junction, Junction),
-	/// A relative path comprising 4 junctions.
-	X4(Junction, Junction, Junction, Junction),
-	/// A relative path comprising 5 junctions.
-	X5(Junction, Junction, Junction, Junction, Junction),
-	/// A relative path comprising 6 junctions.
-	X6(Junction, Junction, Junction, Junction, Junction, Junction),
-	/// A relative path comprising 7 junctions.
-	X7(Junction, Junction, Junction, Junction, Junction, Junction, Junction),
-	/// A relative path comprising 8 junctions.
-	X8(Junction, Junction, Junction, Junction, Junction, Junction, Junction, Junction),
+/// The `parents` field of this struct indicates the number of parent junctions that exist at the
+/// beginning of this `MultiLocation`. A corollary of such a property is that no parent junctions
+/// can be added in the middle or at the end of a `MultiLocation`, thus ensuring well-formedness
+/// of each and every `MultiLocation` that can be constructed.
+#[derive(Clone, Decode, Encode, Eq, PartialEq, Ord, PartialOrd, Debug)]
+pub struct MultiLocation {
+	/// The number of parent junctions at the beginning of this `MultiLocation`.
+	pub(crate) parents: u8,
+	/// The interior (i.e. non-parent) junctions that this `MultiLocation` contains.
+	pub(crate) interior: Junctions,
+}
+
+impl Default for MultiLocation {
+	fn default() -> Self {
+		Self::here()
+	}
 }
 
 /// Maximum number of junctions a `MultiLocation` can contain.
-pub const MAX_MULTILOCATION_LENGTH: usize = 8;
+pub const MAX_MULTILOCATION_LENGTH: usize = 255;
+
+impl MultiLocation {
+	/// Creates a new `MultiLocation`, ensuring that the length of it does not exceed the maximum,
+	/// otherwise returns `Err`.
+	pub fn new(parents: u8, junctions: Junctions) -> result::Result<MultiLocation, ()> {
+		if parents as usize + junctions.len() > MAX_MULTILOCATION_LENGTH {
+			return Err(())
+		}
+		Ok(MultiLocation { parents, interior: junctions })
+	}
+
+	/// Creates a new `MultiLocation` with 0 parents and a `Null` interior.
+	///
+	/// The resulting `MultiLocation` can be interpreted as the "current consensus system".
+	pub const fn here() -> MultiLocation {
+		MultiLocation { parents: 0, interior: Junctions::Null }
+	}
+
+	/// Creates a new `MultiLocation` with the specified number of parents in the `P` const generic
+	/// parameter and a `Null` interior.
+	// #FIXME: Use a where clause to evaluate P for well-formedness once const eval of const
```

Comment:
needs a linked issue.

Reply:
also, what do you mean well-formed-ness? given max-length being 255, `P` is always correct. But your comments imply that otherwise this function has no checks. 

While it is an interesting function, I still don't see a need for it.

Reply:
This function is used in a const eval context, and I can't make the existing `with_parents` work in const, because it returns a `Result`, and unwrapping the result in a const eval context is forbidden.

Since `MAX_MULTILOCATION_LENGTH` is configurable, it's not always going to be `255`, which is why we need a where clause to ensure that P is always less than `MAX_MULTILOCATION_LENGTH`.

Reply:
add a `const_assert!(MAX_MULTILOCATION_LENGTH == u8::MAX)`

And if that fails, we know this function is potentially unsafe. Some comments around the failing test can hint to that.